### PR TITLE
[Proposal] Introduce kvs-schema

### PR DESCRIPTION
**Pull request by rejasupotaro**
# Overview

- Introduced kvs-schema
- Simplified SettingSwitchRowView
- Deleted PrefUtil

# Problem & Solution

Currently, this app retrieves shared preferences through `PrefUtil`. It's a typical way but it has some problems.

## Problem 1: Difficult to guess value types

`PrefUtil` has keys below.

```java
public static final String KEY_CURRENT_LANGUAGE_ID = "current_language_id";
public static final String KEY_NOTIFICATION_SETTING = "notification_setting";
public static final String KEY_HEADS_UP_SETTING = "heads_up_setting";
public static final String KEY_SHOW_LOCAL_TIME = "show_local_time";
```

It looks that `notification_setting` could take json and `show_local_time` could take date. At first glance, I didn't understand what kind of data are stored in this app :confused: 

## Problem 2: Need to declare similar accessors repeatedly

In this way, we need to declare new accessor when we add new value types.

```java
public static void put(Context context, @NonNull String name, boolean value) {
    SharedPreferences.Editor edit = getPref(context).edit();
    edit.putBoolean(name, value);
    edit.apply();
}

public static void put(Context context, @NonNull String name, String value) {
    SharedPreferences.Editor edit = getPref(context).edit();
    edit.putString(name, value);
    edit.apply();
}

public static void put(Context context, @NonNull String name, int value) {
    SharedPreferences.Editor edit = getPref(context).edit();
    edit.putInt(name, value);
    edit.apply();
}

public static void put(Context context, @NonNull String name, float value) {
    SharedPreferences.Editor edit = getPref(context).edit();
    edit.putFloat(name, value);
    edit.apply();
}

public static void put(Context context, @NonNull String name, Set<String> value) {
    SharedPreferences.Editor edit = getPref(context).edit();
    edit.putStringSet(name, value);
    edit.apply();
}

...
```

## Solution

I introduced [kvs-schema](https://github.com/rejasupotaro/kvs-schema) that is a library to manage shared preferences. kvs-schema generates methods below statically using annotation processor.

- `get*`
- `put*`
- `remove*`
- `has*`
- `set*` (for Kotlin)

By this library, we can just focus on key-value management.

# How to introduce kvs-schema

*For the case of existing app*

## Step 1: Add dependencies

Add dependencies to your `build.gradle`.

```groovy
compile 'com.github.rejasupotaro.kvs-schema:library:2.0.0'
apt 'com.github.rejasupotaro.kvs-schema:compiler:2.0.0'
```

## Step 2: Declare schema class.

In this case, `PrefUtil` had keys below.

```java
public static final String KEY_CURRENT_LANGUAGE_ID = "current_language_id";
public static final String KEY_NOTIFICATION_SETTING = "notification_setting";
public static final String KEY_HEADS_UP_SETTING = "heads_up_setting";
public static final String KEY_SHOW_LOCAL_TIME = "show_local_time";
```

Hence, the schema class becomes like below.

```java
@Table("io.github.droidkaigi.confsched_preferences")
public class DefaultPrefsSchema {
    @Key("current_language_id")
    String languageId;
    @Key("notification_setting")
    boolean notificationFlag;
    @Key("heads_up_setting")
    boolean headsUpFlag;
    @Key("show_local_time")
    boolean showLocalTimeFlag;

    private static DefaultPrefs prefs;

    public static synchronized DefaultPrefs get(Context context) {
        if (prefs == null) {
            prefs = new DefaultPrefs(context);
        }
        return prefs;
    }
}
```

Notes :pencil:

- The table name should be the same as the file name. This app already uses **default** shared preferences and its name is `io.github.droidkaigi.confsched_preferences.xml` (package name + `"_preferences"`). So I declared the table name as `"io.github.droidkaigi.confsched_preferences"`.
- The key names should be the same as existing key names as well. On the other hand, you can change field names as you like.
 - I recommend to use `flag` suffix for boolean field like `showLocalTimeFlag`. kvs-schema generates method from field name. If you use `is*` for boolean, the getter method name becomes `getIs*`. It doesn't look cool :confused:
- I declared `DefaultPrefsSchema.get` as a static synchronized method. This method may not be called from many threads at the same time. I guess that the synchronization cost is not so high.

# How to add new key

Just add `@Key("key_name") Type field` to DefaultPrefsSchema.

### CODE REVIEW

**Comment by rejasupotaro** [comment]
@konifar I sent just a proposal PR. I would like to hear your opinion.

**Comment by hotchemi** [comment]
This is what I thought!

File: app/src/main/java/io/github/droidkaigi/confsched/fragment/SettingsFragment.java [hunk 1]
```diff
@@ -55,16 +58,26 @@ public void onAttach(Context context) {
     private void initView() {
         binding.txtLanguage.setText(LocaleUtil.getCurrentLanguage(getActivity()));
         binding.languageSettingsContainer.setOnClickListener(v -> showLanguagesDialog());
-        binding.notificationSwitchRow.init(PrefUtil.KEY_NOTIFICATION_SETTING, true);
-        binding.localTimeSwitchRow.init(PrefUtil.KEY_SHOW_LOCAL_TIME, false);
+
+        boolean shouldNotify = DefaultPrefsSchema.get(getContext()).getNotificationFlag(true);
```
Comment (hotchemi) [inline app/src/main/java/io/github/droidkaigi/confsched/fragment/SettingsFragment.java]: I wonder using getContext causes memory leaks cuz it just returns the Context this fragment is currently associated with. Should we use application context?

refs. https://github.com/rejasupotaro/kvs-schema/blob/06d9eaf3b7b156d7e86b6115b0d2bb68129af77c/library/src/main/java/com/rejasupotaro/android/kvs/PrefsSchema.java#L12

File: app/src/main/java/io/github/droidkaigi/confsched/prefs/DefaultPrefsSchema.java [hunk 2]
```diff
@@ -0,0 +1,27 @@
+package io.github.droidkaigi.confsched.prefs;
+
+import android.content.Context;
+
+import com.rejasupotaro.android.kvs.annotations.Key;
+import com.rejasupotaro.android.kvs.annotations.Table;
+
+@Table("io.github.droidkaigi.confsched_preferences")
+public class DefaultPrefsSchema {
```
Comment (hotchemi) [inline app/src/main/java/io/github/droidkaigi/confsched/prefs/DefaultPrefsSchema.java]: Is it ok that this class is not abstract?

**Comment by hotchemi** [comment]
Almost awesome! This app will be a showcase of Japanese libraries.

File: app/src/main/java/io/github/droidkaigi/confsched/fragment/SettingsFragment.java (hunk 1 above)
Comment (rejasupotaro) [inline app/src/main/java/io/github/droidkaigi/confsched/fragment/SettingsFragment.java]: I think we don't have to use application context. Generally speaking, memory leak could occur when given context is used in different lifetime. Meanwhile in kvs-schema, `DefaultPrefsSchema.get` runs synchronoushly and this library doesn't hold a reference of context.

File: app/src/main/java/io/github/droidkaigi/confsched/prefs/DefaultPrefsSchema.java (hunk 2 above)
Comment (rejasupotaro) [inline app/src/main/java/io/github/droidkaigi/confsched/prefs/DefaultPrefsSchema.java]: Good point! I updated it in https://github.com/rejasupotaro/droidkaigi2016/commit/655310dd4573ee9ee94b22afe0051595553dbcd8

**Comment by sys1yagi** [comment]
:heart_eyes:

File: app/src/main/java/io/github/droidkaigi/confsched/prefs/DefaultPrefsSchema.java (hunk 2 above)
Comment (hotchemi) [inline app/src/main/java/io/github/droidkaigi/confsched/prefs/DefaultPrefsSchema.java]: Thx!

File: app/src/main/java/io/github/droidkaigi/confsched/fragment/SettingsFragment.java (hunk 1 above)
Comment (hotchemi) [inline app/src/main/java/io/github/droidkaigi/confsched/fragment/SettingsFragment.java]: ok!

**Comment by konifar** [comment]
:eyes:

**Comment by konifar** [comment]
Thanks for your cool solution and detailed description!
LGTM!

**Pull request merged**